Plausible/DeriveArbitrary.lean

```diff
@@ -0,0 +1,281 @@
+      let pureGen ← `($pureFn $ctorIdent)
+      nonRecursiveGenerators := nonRecursiveGenerators.push pureGen
+    else
+      -- Add all the freshened names for the constructor to the local context,
+      -- then produce the body of the sub-generator
+      let (generatorBody, ctorIsRecursive) ← liftTermElabM $
+        withLocalDeclsDND freshArgNamesTypes (fun _ => do
+          let mut doElems := #[]
+
+          -- Determine whether the constructor has any recursive arguments
+          let ctorIsRecursive ← isConstructorRecursive targetTypeName ctorName
+          if !ctorIsRecursive then
+            -- Call `arbitrary` to generate a random value for each of the arguments
+            for freshIdent in freshArgIdents do
+              let bindExpr ← mkLetBind freshIdent #[arbitraryFn]
+              doElems := doElems.push bindExpr
+          else
+            -- For recursive constructors, we need to examine each argument to see which of them require
+            -- recursive calls to the generator
+            for (freshIdent, argType) in freshArgIdentsTypes do
+              -- If the argument's type is the same as the target type,
+              -- produce a recursive call to the generator using `aux_arb`,
+              -- otherwise generate a value using `arbitrary`
+              let bindExpr ←
+                if argType.getAppFn.constName == targetTypeName then
+                  mkLetBind freshIdent #[auxArbFn, freshSize']
+                else
+                  mkLetBind freshIdent #[arbitraryFn]
+              doElems := doElems.push bindExpr
+
+          -- Create an expression `return C x1 ... xn` at the end of the generator, where
+          -- `C` is the constructor name and the `xi` are the generated values for the args
+          let pureExpr ← `(doElem| return $ctorIdent $freshArgIdents*)
+          doElems := doElems.push pureExpr
+
+          -- Put the body of the generator together
+          let generatorBody ← mkDoBlock doElems
+          pure (generatorBody, ctorIsRecursive))
+
+      if !ctorIsRecursive then
+        nonRecursiveGenerators := nonRecursiveGenerators.push generatorBody
+      else
+        recursiveGenerators := recursiveGenerators.push generatorBody
+
+  -- Just use the first non-recursive generator as the default generator
+  let defaultGenerator := nonRecursiveGenerators[0]!
+
+  -- Turn each generator into a thunked function and associate each generator with its weight
+  -- (1 for non-recursive generators, `.succ size'` for recursive generators)
+  let thunkedNonRecursiveGenerators ←
+    Array.mapM (fun generatorBody => `($thunkGenFn (fun _ => $generatorBody))) nonRecursiveGenerators
+
+  let mut weightedThunkedNonRecursiveGens := #[]
+  for thunkedGen in thunkedNonRecursiveGenerators do
+    let thunkedGen ← `((1, $thunkedGen))
+    weightedThunkedNonRecursiveGens := weightedThunkedNonRecursiveGens.push thunkedGen
```

Comment:
This can be moved up when the arrays are being constructed.

For example
```
let gen ← ``((1, Gen.thunkGen (fun _ => $generatorBody)))
nonRecursiveGenerators := nonRecursiveGenerators.push entry
```

(These loops can be merged by the way.)